add elastic training blog (#78)
kubeflow/blog: _posts/2021-03-15-elastic-training.md

- When collocating training tasks with other workloads (with higher priority), the resources demand fluctuates as the request for these other workloads may change periodically. This unbalance of resources availability throws cold water on the idea of using hybrid-deployment to maximize resource utilization.
 
## Elastic Training
 
Researchers and engineers proposed **Elastic Training** as the key to solve the puzzle.
 
Traditionally, the resource configuration for a distributed training job is fixed. Elastic training breaks this rule and enables users to change the number of instances participating in a distributed training job, bringing the following benefits to clusters with distributed training jobs:
 
1. Fault Tolerance: any worker instance can fail as long as at least one is surviving.
2. Resources Utilization: when the resources stress piles, the cluster is able to reduce the replicas of workloads with lower priority (distributed training workloads), releasing resource to other workloads (such as prediction service), ensuring SLA for business; after resources released from workloads, elastic training job is able to absorb these resource by scaling up workload replicas.

> 3. Training on Cloud: there is a type of resource on the cloud that is called "spot" or "preemptible" instances; it comes with unexptected low price tags but may be retrieved after guaranteed hour expires.

Reviewer: ```suggestion
3. Training on Cloud: there is a type of resource on the cloud that is called "spot" or "preemptible" instances; it comes with unexpected low price tags but may be retrieved after guaranteed hour expires.
```

 
## Elastic Horovod
 
As the major player in distributed training framework, Horovod v0.20.0 offers its solution to elastic training, [Elastic Horovod](https://horovod.readthedocs.io/en/stable/elastic_include.html). Here we quotes the architecture differences between Elastic Horovod and existing Horovod from [RFC Elastic Horovod](https://docs.google.com/document/d/15ZoHA5AeSI_boeyIBapg9WPXKrYXMRvPytPzQWTCTn4/edit#):

<img src="/images/2021-03-15-elastic-training/horovod-elastic.png" width="" alt="alt_text" title="">
 
* All collective operations are coordinated within a hvd.elastic.run function.
* State is synchronized between workers before the user’s training function is executed.
* Worker failure or worker added events will result in a reset event on other workers.
* Reset events act as barriers to:
  * Determine whether the job should continue based on worker exit codes.
  * Blacklist failing hosts.
  * Launch workers on new hosts.
  * Update the rank information on existing workers.
* State is synchronized following a reset event.
 

> When launching an elastic training job, `horovodrun` requires a `discover_hosts.sh` script to detect available hosts and slots in real time. (In the following section, we refer this script as `discover_hosts.sh`. Nevertheless the script needs not to be named as `discover_hosts.sh`.)

Reviewer: Is there any link that we can add here for the `discover_hosts.sh` script ?

Author: This example (https://horovod.readthedocs.io/en/stable/elastic_include.html#running-with-horovodrun) includes what the `discover_hosts.sh` is expected to return.

Reviewer: @zw0610 It's good. Let's include it in the blog.

 
## Elastic Horovod on Kubernetes
 
[MPI-Operator](https://github.com/kubeflow/mpi-operator) is designed to deploy Horovod jobs on Kubernetes. While the operator releases multiple versions, the general idea stays unchanged. It includes:

<img src="/images/2021-03-15-elastic-training/mpi-operator.png" width="" alt="alt_text" title="">
 

> 1. MPIJob Controller creates a launcher pod and worker pods according to the replicas configuration in MPIJobs

Reviewer: Should we try to be consistent between graph numeration and these steps ?

Author: Since we do not want to re-iterate the mpi-operator proposal in this blog, I changed the numeration to bullet so readers should be not confused. What do you think?

Reviewer: Thank you, I think it is better since it will not confuse reader.

 
## Demo
 
We present a demo to show how to operate an Elastic Horovod job with MPI Operator. 
 
```bash
bash-5.0$ kubectl create -f ./tensorflow-mnist-elastic.yaml
mpijob.kubeflow.org/tensorflow-mnist-elastic 
createdbash-5.0$ kubectl get po
NAME    READY   STATUS    RESTARTS  AGE
tensorflow-mnist-elastic-launcher   1/1     Running   0          14s
tensorflow-mnist-elastic-worker-0   1/1     Running   0          14s
tensorflow-mnist-elastic-worker-1   1/1     Running   0          14s
```
 
The job is created with two workers. After the training begins, we change `MPIJob.Spec.MPIReplicaSpecs["Worker"].Replicas` to **3**, adding another worker. Let's check how the `discover_hosts.sh` changes:
 
```bash
bash-5.0$ kubectl exec tensorflow-mnist-elastic-launcher -- /etc/mpi/discover_hosts.sh
tensorflow-mnist-elastic-worker-0:1
tensorflow-mnist-elastic-worker-1:1

> bash-5.0$ kubectl edit mpijob/tensorflow-mnist-elastic

Reviewer: Maybe we can use `kubectl patch` to be more explicit here, what do you think ?

Author: good idea. I updated the script in the new commit.

tensorflow-mnist-elastic-worker-2   1/1     Terminating   0          2m21s
```
 
The elastic training persists.

```bash
Thu Mar 11 01:53:18 2021[1]<stdout>:Step #40    Loss: 0.284265
Thu Mar 11 01:53:18 2021[0]<stdout>:Step #40    Loss: 0.259497
Thu Mar 11 01:53:18 2021[2]<stdout>:Step #40    Loss: 0.229993
Thu Mar 11 01:54:27 2021[2]<stderr>:command terminated with exit code 137
Process 2 exit with status code 137.
Thu Mar 11 01:54:27 2021[0]<stderr>:command terminated with exit code 137
Process 0 exit with status code 137.
Thu Mar 11 01:54:57 2021[1]<stderr>:[2021-03-11 01:54:57.532928: E /tmp/pip-install-2jy0u7mn/horovod/horovod/common/operations.cc:525] Horovod background loop uncaught exception: [/tmp/pip-install-2jy0u7mn/horovod/third_party/compatible_gloo/gloo/transport/tcp/pair.cc:575] Connection closed by peer [10.244.2.27]:54432
WARNING:root:blacklist failing host: tensorflow-mnist-elastic-worker-2
WARNING:root:blacklist failing host: tensorflow-mnist-elastic-worker-1
Thu Mar 11 01:54:58 2021[1]<stdout>:Step #50    Loss: 0.207741
Thu Mar 11 01:55:00 2021[1]<stdout>:Step #60    Loss: 0.119361
Thu Mar 11 01:55:02 2021[1]<stdout>:Step #70    Loss: 0.131966
```
 

> As we can see, Elastic Horovod on MPI-Operator now supports tweaking worker replicas dynamically. As a future work, we aim to support HorizontalPodAutocaler to MPIJob as well as other features like designated worker deletion.

Reviewer: Should we add link here? 
```suggestion
As we can see, Elastic Horovod on MPI-Operator now supports tweaking worker replicas dynamically. As a future work, we aim to support  [`Horizontal Pod Autoscaler`](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/) to MPIJob as well as other features like designated worker deletion.
```

---
toc: true
layout: post
categories: [elastic training, operators]

> title: "Elastic Training with MPI Operator and Practise"

Reviewer 2: Practise -> Practice

Author: fixed

comments: true
author: "<a href='https://www.linkedin.com/in/gaocegege/'>Ce Gao</a>, <a href='https://www.linkedin.com/in/wang-zhang'>Wang Zhang</a>"
---
 

> With increase in the size of dataset and deep learning models, distributed training emerges as the mainstream approach for training neural network models in industry. While it is feasible now to launch a massive distributed training job on Kubernetes with Kubeflow, advanced features like elastic workload and other cost mitigation approaches remain leashed when we talk about deep learning jobs on Kubernetes.

Reviewer 2: With increasing of the dataset and deep learning models ..?

To address issues on cost and resource utilization, the [TKE (Tencent Kubernetes Engine)](https://intl.cloud.tencent.com/product/tke) AI team designs and implements **Elastic Training** in Kubeflow community.
 
Here we present how the elastic training is performed on Kubernetes. Validated with experiments under circumstances, elastic training lowers cost for distributed training on cloud.
 
## Background
 

> Let's first recap training deep learning models. When we talk about 'training', it refers generally to iteratively optimizing parameters in a neural network model with its gradient descent. Accelerated with GPUs, the training can speed up for 10-100 times.

Reviewer 2: `training`? change ' to `

Author: single quotation seems fine as it's not referring code script.

 
## Elastic Horovod
 
As the major player in distributed training framework, Horovod v0.20.0 offers its solution to elastic training, [Elastic Horovod](https://horovod.readthedocs.io/en/stable/elastic_include.html). Here we quotes the architecture differences between Elastic Horovod and existing Horovod from [RFC Elastic Horovod](https://docs.google.com/document/d/15ZoHA5AeSI_boeyIBapg9WPXKrYXMRvPytPzQWTCTn4/edit#):

<img src="/images/2021-03-15-elastic-training/horovod-elastic.png" width="" alt="alt_text" title="">
 
* All collective operations are coordinated within a hvd.elastic.run function.
* State is synchronized between workers before the user’s training function is executed.
* Worker failure or worker added events will result in a reset event on other workers.
* Reset events act as barriers to:
  * Determine whether the job should continue based on worker exit codes.
  * Blacklist failing hosts.
  * Launch workers on new hosts.
  * Update the rank information on existing workers.
* State is synchronized following a reset event.
 

> When launching an elastic training job, `horovodrun` requires a `discover_hosts.sh` script to detect available hosts and slots in real time. (In the following section, we refer this script as `discover_hosts.sh`. Nevertheless the script needs not to be named as `discover_hosts.sh`. An example of `discover_hosts.sh` can be found [here](https://horovod.readthedocs.io/en/stable/elastic_include.html#running-with-horovodrun).)

Reviewer: ```suggestion
When launching an elastic training job, `horovodrun` requires a `discover_hosts.sh` script to detect available hosts and slots in real time. (In the following section, we refer this script as `discover_hosts.sh`. Nevertheless the script needs not to be named as `discover_hosts.sh`. An example of `discover_hosts.sh` can be found [here](https://horovod.readthedocs.io/en/stable/elastic_include.html#running-with-horovodrun).
```

Author: There is a parentheses before "In the following section", which makes a pair with the one at the end of the paragraph.
Are you suggesting remove this pair of parentheses?

Reviewer: Oh, just missed that.
Yes, I think we can remove it.
